core/trino-main/src/main/java/io/trino/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -370,7 +371,9 @@ private synchronized ClientBuffer getBuffer(OutputBufferId id)
         // NOTE: buffers are allowed to be created in the FINISHED state because destroy() can move to the finished state
         // without a clean "no-more-buffers" message from the scheduler.  This happens with limit queries and is ok because
         // the buffer will be immediately destroyed.
-        checkState(stateMachine.getState().canAddBuffers() || !outputBuffers.isNoMoreBufferIds(), "No more buffers already set");
+        BufferState state = stateMachine.getState();
+        // buffer may become aborted while the final output buffers are being set
+        checkState(state == ABORTED || state.canAddBuffers() || !outputBuffers.isNoMoreBufferIds(), "No more buffers already set");
 
         // NOTE: buffers are allowed to be created before they are explicitly declared by setOutputBuffers
         // When no-more-buffers is set, we verify that all created buffers have been declared
```

Comment:
reuse state below `        if (stateMachine.getState() == FINISHED) {`? Is there a change for race condition?

Reply:
In theory there shouldn't be a race condition, as `getBuffer` is synchronized, and `safeGetBuffersSnapshot()` used by `finish` is also synchronized. But somehow I'm still reluctant of changing semantics.

Reply:
The reason it looks like a race is because ` getBuffer` is synchronzied, but buffer state machine isn't, so I though that state might change while `getBuffer` is progressing.

Neither `io.trino.execution.buffer.ArbitraryOutputBuffer#setNoMorePages` or `io.trino.execution.buffer.ArbitraryOutputBuffer#destroy()` is synchronized